pkg/controller/elasticsearch/reconcile/status.go

```diff
@@ -0,0 +1,304 @@
+	})
+}
+
+// -- Upscale status
+
+type UpscaleReporter struct {
+	// Expected nodes to be upscaled
+	nodes map[string]esv1.NewNode
+}
+
+func (u *UpscaleReporter) RecordNewNodes(nodes []string) {
+	if u == nil {
+		return
+	}
+	if u.nodes == nil {
+		u.nodes = make(map[string]esv1.NewNode, len(nodes))
+	}
+	if nodes == nil {
+		nodes = []string{}
+	}
+	for _, node := range nodes {
+		newNode := u.nodes[node]
+		newNode.Name = node
+		newNode.Status = esv1.NewNodePending
+		u.nodes[node] = newNode
+	}
+}
+
+// UpdateNodesStatuses the status and the message fields for a set of nodes.
+func (u *UpscaleReporter) UpdateNodesStatuses(status esv1.NewNodeStatus, statefulSetName, message string, minOrdinal, maxOrdinal int32) {
+	if u == nil {
+		return
+	}
+	if u.nodes == nil {
+		u.nodes = make(map[string]esv1.NewNode)
+	}
+	for ord := minOrdinal - 1; ord < maxOrdinal; ord++ {
+		podName := sset.PodName(statefulSetName, ord)
+		newNode := u.nodes[podName]
+		newNode.Status = status
+		newNode.Message = pointer.String(message)
+		u.nodes[podName] = newNode
+	}
+}
+
+func (u *UpscaleReporter) HasPendingNewNodes() bool {
+	if u == nil {
+		return false
+	}
+	return len(u.nodes) > 0
+}
+
+func (u *UpscaleReporter) Merge(other esv1.UpscaleOperation) esv1.UpscaleOperation {
+	upscaleOperation := other.DeepCopy()
+	if u == nil {
+		return *upscaleOperation
+	}
+	nodes := make([]esv1.NewNode, 0, len(u.nodes))
+	for name, node := range u.nodes {
+		nodes = append(nodes, esv1.NewNode{
+			Name:    name,
+			Status:  node.Status,
+			Message: node.Message,
+		})
+	}
+	// Sort for stable comparison
+	sort.Slice(nodes, func(i, j int) bool {
+		return nodes[i].Name < nodes[j].Name
+	})
+	if (u.nodes != nil && !reflect.DeepEqual(nodes, other.Nodes)) || upscaleOperation.Nodes == nil {
+		upscaleOperation.Nodes = nodes
+		upscaleOperation.LastUpdatedTime = metav1.Now()
+	}
+	return *upscaleOperation
+}
+
+// -- Upgrade status
+
+type UpgradeReporter struct {
+	// Expected nodes to be upgraded, key is node name
+	nodes map[string]esv1.UpgradedNode
+}
+
+func (u *UpgradeReporter) RecordNodesToBeUpgraded(nodes []string) {
+	u.RecordNodesToBeUpgradedWithMessage(nodes, "")
+}
+
+func (u *UpgradeReporter) RecordNodesToBeUpgradedWithMessage(nodes []string, message string) {
+	if u == nil {
+		return
+	}
+	if u.nodes == nil {
+		u.nodes = make(map[string]esv1.UpgradedNode, len(nodes))
+	}
+	for _, node := range nodes {
+		upgradedNode := u.nodes[node]
+		upgradedNode.Name = node
+		upgradedNode.Status = "PENDING"
+		if len(message) > 0 {
+			upgradedNode.Message = pointer.String(message)
+		}
+		u.nodes[node] = upgradedNode
+	}
+}
+
+func (u *UpgradeReporter) RecordDeletedNode(node, message string) {
+	if u == nil {
+		return
+	}
+	if u.nodes == nil {
+		u.nodes = make(map[string]esv1.UpgradedNode)
+	}
+	upgradedNode := u.nodes[node]
+	upgradedNode.Name = node
+	upgradedNode.Status = "DELETED"
+	upgradedNode.Message = pointer.String(message)
+	u.nodes[node] = upgradedNode
```

Comment:
> We could have just one private helper method that the other methods `RecordDeletedNode` and `RecordNodeToBeUpdatedWithMessage` delegate to, to avoid some duplication of initialisation etc.

👍 https://github.com/elastic/cloud-on-k8s/commit/093ee9e9a47f344dd18cab3e5be0077594e18028